app/lua/lstate.h

```diff
@@ -82,7 +82,7 @@ typedef struct global_State {
   Mbuffer buff;  /* temporary buffer for string concatentation */
   lu_mem GCthreshold;
   lu_mem totalbytes;  /* number of bytes currently allocated */
-  lu_mem memlimit;  /* maximum number of bytes that can be allocated, 0 = no limit. */
+  l_mem memlimit;  /* maximum number of bytes that can be allocated, 0 = no limit. <0 used with EGC_ON_MEM_LIMIT when free heap falls below -memlimit */
```

Comment:
Johny, you can't overload the g->memlimit this way without going through all of the places in the core which use it.  For example, it is also used in `lapi.c` and `lvm.c`

Reply:
Through the wonders of int -> unsigned conversion, I can :)

In `lapi.c` the only use is through `collectgarbage("setmemlimit", newlimit)`. I did not change it because I see no reason to. Using a negative memlimit is an extension available through `node.egc.setmode()` _only_. Setting a regular *hard* memlimit via `collectgarbage()` is still possible (and if you want, you could force negative values through it, but no sane value on the ESP8266 would interfere).

In `lvm.c` I left it alone because a negative memlimit expressed as an unsigned is effectively the same as `MAX_SIZET` on our tiny architecture.

Reply:
@jmattsson Johny, IMO, you aren't fully correct in your assertion.  `G(L)->memlimit` is a single global `uint` variable that was introduced in the `eLua` patch but is used in more than just the EGC threshold calcs. 
-  In `lvm.c:luaV_concat()`, the maximum string length allowed is the lesser of `MAX_SIZET` and  `G(L)->memlimit`.  this this is negative then we will be checking the string length against a negative value and always throw a **"string length overflow"** error. This logic needs fixing or dropping.
-  the GC `setmemlimit` options is an eLua/EGC addition and is an alternative API for setting this ECG variable.  `memlimit`isn;t used by the regular GC, only the EGC add-on.

Reply:
However, the negative `G(L)->memlimit` gets assigned into an `unsigned`, at which point it becomes a very large positive number, which on the ESP8266 is way above the available memory that the result is effectively the same as `MAX_SIZET`. Try it for yourself if you don't believe me :)

Reply:
OK, I agree. the size of a signed vs unsigned comparison or coerced to an unsigned comparison, but looking at this code I really have to wonder what the rationale for the EGC patch change was.  Why include a bounds check to ensure that the size of a concat string is < MAXUINT - a tiny bit?

However, that is your problem, so I accept your comment.